### ISSUE TITLE
request: add xkb_keymap_layout_get_name() variant returning 2-letter layout

### ISSUE THREAD

**Issue by bam80** [xkbcommon/libxkbcommon#192]
It was asked https://github.com/xkbcommon/libxkbcommon/issues/20#issuecomment-690819358 but no answer there..

**Comment by bam80** [comment]
I'm stuck on it, could you please pay attention? @bluetech @fooishbar 
Without getting the short 2-letter layout name from `libxkbcommon` I don't really see the way to implement layout indicator properly for KDE Plasma. I know I could parse the registry .XML but it would make code ugly and should not be needed for this purpose. I guess `libxkbcommon` already 'knows' these shot names, shall we just expose it?

**Comment by bluetech** [comment]
Unfortunately libxkbcommon doesn't know these names. The names map into keymaps but this mapping is not reversible in general.

Could you explain your use case in more detail? I understand it's a keyboard layout indicator, but where does the keymap come from? Where is it configured? How is it generated (Wayland protocol, x11 protocol, locally)? etc.

**Comment by bam80** [comment]
Thank you for a quick reply Ran!

Currently I implementing it for KWin Wayland compositor, but we are targeting to make it protocol-agnostic.
KDE Plasma has Keyboard configuration module where it parses the .XML registry to configure layouts etc.:
https://lxr.kde.org/source/kde/workspace/plasma-desktop/kcms/keyboard/xkb_rules.cpp#0179
https://lxr.kde.org/source/kde/workspace/plasma-desktop/kcms/keyboard/README

But also there is `Configure Layouts` tick where all that configuration on Plasma side could be essentially disabled, leaving the user with system(user?)-default.
In that case, I believe configuration can come from all the config paths `libxkbcommon` parses and even from environment variables (XKB_DEFAULT_LAYOUT).

Please ask if you need more information.

**Comment by bluetech** [comment]
OK so if I understand correctly, extracting the short name is needed on the compositor side, not the client side, right? And then the compositor exposes this info to the layout indicator in some fashion?

Could you point me to where the `xkb_keymap` is created?

**Comment by bam80** [comment]
> OK so if I understand correctly, extracting the short name is needed on the compositor side, not the client side, right? And then the compositor exposes this info to the layout indicator in some fashion?
> 
Right, it exposes it via DBus signal, passing long descriptive layout name currently.

> Could you point me to where the `xkb_keymap` is created?

https://lxr.kde.org/source/kde/workspace/kwin/xkb.cpp#0123
https://lxr.kde.org/ident?_i=xkb_keymap&_remember=1

**Comment by bluetech** [comment]
Thanks for the pointer.

So as I said before, a function such as `xkb_keymap_layout_get_name()` is not generally possible, the way the XKB stack is laid out. The lowest level, an XKB layout/"group"/"symbols", which is what `xkb_keymap` represents, only has a long name. The short name only serves to find the layout (first through the rules file, then through the filesystem). So even if we wanted to add it it would be thwarted by issues like:

- Not all input formats even use the short name -- for example, if you look at the output of `xkbcli compile-keymap --layout us` you can see that the short name doesn't appear. Wayland clients for example use such an input format (serialized from the compositor). So the function would really only be ~possible for keymaps created using `xkb_keymap_new_from_names`.

- XKB layout names can actually be pretty weird, for example `us+ru+fr` is a perfectly legal layout name (note the `+` rather than the usual `,`).

- It is possible for multiple short names to map to the same long name, so it's only possible to go one way.

- For some layouts a short name does not exist.

Given that, I believe the best way for you to implement this feature is to keep the `layoutList` which you use to generate the `xkb_keymap` along with the `xkb_keymap`, and use it to get the short name, based on the layout index. Or you can go one step further and look at the metadata from which the layoutList was chosen. Is this possible for you to do?

Also cc @whot who has recently done some similar work on this in gnome ([here](https://gitlab.gnome.org/GNOME/gnome-desktop/-/blob/master/libgnome-desktop/gnome-xkb-info.c)).

**Comment by bam80** [comment]
Thank you for a comprehensive answer Ran!

>     * XKB layout names can actually be pretty weird, for example `us+ru+fr` is a perfectly legal layout name
> 
That is why I'm thinking to make them editable in the future..
Still, I'm interesting what is use case for such layout and what would be it's long name?

> Given that, I believe the best way for you to implement this feature is to keep the `layoutList` which you use to generate the `xkb_keymap` along with the `xkb_keymap`, and use it to get the short name, based on the layout index. 

Yep, that would work!

> Or you can go one step further and look at the metadata from which the layoutList was chosen. Is this possible for you to do?
> 

That is what I have yet to find out for myself :)
Apparently something similar is already done in Plasma's Keyboard configuration, deducing https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes or what user have entered, but that result seems is not saved in the config.. I could change that and use processed values in compositor instead.

Well, even if the issue can be considered as solved for this particular case, overall it feels to me we need document it better somewhere. It is not obvious. Maybe leave some notes in [xkb_keymap_layout_get_name()](https://xkbcommon.org/doc/current/group__components.html#ga63b85a1431caed99af85e19ff50131f6) description itself?

**Comment by whot** [comment]
>  Still, I'm interesting what is use case for such layout and what would be it's long name?

use case: we're dealing with languages here, so everything is weird :)

think of the layout names vs the keymap as restaurant menu item vs a meal. It's easy to go from "spaghetti bolognese" (the long name) to a meal (the keymap) but it's harder to look at some meal served and identify it without knowing the list of menu items, especially for nonstandard cases. Most people configure via RMLVO (which is what you do in `Xkb::loadKeymapFromConfig()`) but that's not the only use-case.

How RMLVO is converted to the keymap is implementation-defined in the xkeyboard-config rules files, I recently wrote a [set of posts](https://who-t.blogspot.com/2020/09/no-user-specific-xkb-configuration-in-x.html) about this (link goes to part 5 but has the links to parts 1-4). And, sadly enough, the short names you're looking at are often nothing more than filenames required by xkeyboard-config (e.g. "fr" will map to "symbols/fr"). However, xkeyboard-config has the xml files with extra data and
with libxkbregistry you can get the ISO639 codes for the RMLVO configuration you currently use. You don't have to parse the xml directly anymore.

So basically, your task requires a bit of rethinking to line yourself up with the one-way conversion of keymaps:
- the user configures keyboard layouts via RMLVO in some config interface. That's already the case and has been for years.
- the base assumption is that RMLVO produces the keymap and that keymap is never altered, thus the RMVLO is the only configuration data you ever need to care about
- the RMLVO you used to generate the keymap is stored and accessible to the other tools
- you can use that RMLVO via libxkbregistry to look up ISO639 and ISO 3166 codes which you can then present to the user

niche cases like directly loading a keymap without going through RMLVO first are possible but a plate of random stew won't give you a proper meal name, there simply may not be one.

**Comment by bam80** [comment]
Hi Peter, thank you for your input.

> the RMLVO you used to generate the keymap is stored and accessible to the other tools

I'm not sure what does it mean. RMLVO is stored in Plasma's config and used by it only, isn't it?
I'm not aware of any portable format for RMLVO. Or you mean something else?

**Comment by whot** [comment]
if that's already the case, then implementing that step is easy ;)

**Comment by bam80** [comment]
filed #193 instead. Thanks anyone.

**Comment by bluetech** [comment]
BTW, another thing you might want to pay attention to is the variant. A user might have for example two layouts `us` and `us(intl)`. Presumably the layout indicator would want to distinguish them in some way. Gnome went with just `en₁`  and `en₂` for this.

**Comment by bam80** [comment]
That is another argument for have that layout names editable. Actually, Plasma already have this working for X11 applet, just need to expose that functionality to Wayland compositor.
What comes to mind first is just store these processed values (ISO 639 codes or what user has edited) to config along with `true` shot layout names for the library. Compositor will use the processed values to display.

**Comment by bam80** [comment]
For the reference, here is the current KWin's MR for the feature as discussed above:
[Draft: [1/2] feat: expose keyboard layout Short Name via DBus ](https://invent.kde.org/plasma/kwin/-/merge_requests/428)
